src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(mprotect), 60);
+	if (rc < 0)
+		goto fail;
+
+	/* mmap */
+	struct mmap_struct_t {
+		int prot;
+		int flags;
+	};
+	const struct mmap_struct_t mmap_anon_struct[4] = {
+		{ PROT_NONE,            MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS               },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK     },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS     }
+	};
+	for (i = 0; i < 4; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mmap), 4,
+			SCMP_A2(SCMP_CMP_EQ, mmap_anon_struct[i].prot),
+			SCMP_A3(SCMP_CMP_EQ, mmap_anon_struct[i].flags),
+			SCMP_A4(SCMP_CMP_EQ, 0xffffffff),
+			SCMP_A5(SCMP_CMP_EQ, 0));
+		if (rc < 0)
+			goto fail;
+	}
+
+	const struct mmap_struct_t mmap_noanon_struct[3] = {
+		{ PROT_READ,           MAP_PRIVATE               },
+		{ PROT_READ,           MAP_SHARED                },
+		{ PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE }
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mmap), 4,
+			SCMP_A2(SCMP_CMP_EQ, mmap_noanon_struct[i].prot),
+			SCMP_A3(SCMP_CMP_EQ, mmap_noanon_struct[i].flags),
+			SCMP_A4(SCMP_CMP_NE, 0xffffffff),
+			SCMP_A5(SCMP_CMP_EQ, 0));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mmap), 4,
+		SCMP_A2(SCMP_CMP_EQ, PROT_READ|PROT_WRITE),
+		SCMP_A3(SCMP_CMP_EQ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE),
+		SCMP_A4(SCMP_CMP_NE, 0xffffffff),
+		SCMP_A5(SCMP_CMP_NE, 0)); /* NE 0, so can't be in the struct */
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(mmap), 60);
+	if (rc < 0)
+		goto fail;
+
+	/* madvise */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(madvise), 2,
+		SCMP_A1(SCMP_CMP_EQ, 8368128),
+		SCMP_A2(SCMP_CMP_EQ, MADV_DONTNEED));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(madvise), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* socket */
+	struct socket_struct_t {
+		int domain;
+		int type;
+		int proto;
+	};
+	const struct socket_struct_t socket_struct[2] = {
+		{ PF_INET, SOCK_STREAM, IPPROTO_TCP },
+		{ PF_INET, SOCK_STREAM, IPPROTO_IP  }
+	};
+	for (i = 0; i < 2; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(socket), 3,
+			SCMP_A0(SCMP_CMP_EQ, socket_struct[i].domain),
+			SCMP_A1(SCMP_CMP_EQ, socket_struct[i].type),
+			SCMP_A2(SCMP_CMP_EQ, socket_struct[i].proto));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(socket), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* connect */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(connect), 2,
+		SCMP_A0(SCMP_CMP_GE, 4),
+		SCMP_A2(SCMP_CMP_EQ, 16));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(connect), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* getpeername */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(getpeername), 1,
+		SCMP_A0(SCMP_CMP_GE, 4));
```

Comment:
Then I agree. For these at least, hardcoded fds seem like a pain. Perhaps if we have seccomp wait until irssi connects to the network, we could whitelist the fds that are actually in use though, something like `enable_seccomp_sandbox(network_fd, some_other_fd, yet_another_fd)`.